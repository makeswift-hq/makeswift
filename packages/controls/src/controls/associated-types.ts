import { ControlDefinition } from './definition'

export type ControlType<D> =
  D extends ControlDefinition<infer ControlType, any, any, any, any, any>
    ? ControlType
    : never

export type ConfigType<D> =
  D extends ControlDefinition<any, infer Config, any, any, any, any>
    ? Config
    : never

export type DataType<D> =
  D extends ControlDefinition<any, any, infer DataType, any, any, any>
    ? DataType
    : never

export type ValueType<D> =
  D extends ControlDefinition<any, any, any, infer ValueType, any, any>
    ? ValueType
    : never

export type ResolvedValueType<D> =
  D extends ControlDefinition<any, any, any, any, infer ResolvedValueType, any>
    ? ResolvedValueType
    : never

export type InstanceType<D> =
  D extends ControlDefinition<any, any, any, any, any, infer InstanceType>
    ? InstanceType
    : never
