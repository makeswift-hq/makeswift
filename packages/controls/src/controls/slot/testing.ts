import { SlotDefinition } from './slot'

export const renderedNode = Symbol('rendered node')

export type RenderedNode = typeof renderedNode

class Definition extends SlotDefinition<RenderedNode> {}

export function Slot(): Definition {
  return new Definition()
}

export { Definition as SlotDefinition }
