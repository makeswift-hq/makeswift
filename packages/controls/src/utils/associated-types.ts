export type WithAssociatedTypes<T> = {
  readonly __associated_types__?: T
}

export type AssociatedTypes<T> = T extends {
  __associated_types__?: infer Types
}
  ? Types extends Record<string, unknown>
    ? Types
    : never
  : never

export type HasAssociatedTypes<T> = T extends {
  __associated_types__?: infer Types
}
  ? Types extends Record<string, unknown>
    ? true
    : false
  : false

export type AssociatedType<
  T,
  K extends keyof AssociatedTypes<T>,
> = AssociatedTypes<T>[K]
