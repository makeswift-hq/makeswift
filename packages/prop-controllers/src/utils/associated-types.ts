export type WithAssociatedTypes<T> = {
  readonly __associated_types__?: T
}

export type AssociatedTypes<T> = T extends {
  __associated_types__?: infer Types
}
  ? Types extends Record<string, unknown>
    ? Types
    : never
  : never

export type AssociatedType<
  T,
  K extends keyof AssociatedTypes<T>,
> = AssociatedTypes<T>[K]
