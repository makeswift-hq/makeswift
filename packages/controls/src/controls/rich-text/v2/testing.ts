import { z } from 'zod'

import { type SerializedRecord } from '../../../serialization'
import { type DataType } from '../../associated-types'
import { serialize } from '../../definition'
import {
  ControlInstance,
  DefaultControlInstance,
  type SendMessage,
} from '../../instance'

import { RichTextPluginControl } from './plugin'
import { RichTextDefinition } from './rich-text'

export const renderedNode = Symbol('rendered node')

export type RenderedNode = typeof renderedNode

type UserConfig = z.infer<typeof Definition.schema.userConfig>

class Definition extends RichTextDefinition<RenderedNode> {
  createInstance(sendMessage: SendMessage<any>): ControlInstance<any> {
    return new DefaultControlInstance(sendMessage)
  }

  serialize(): [SerializedRecord, Transferable[]] {
    return serialize(this.config, {
      type: Definition.type,
    })
  }

  toText(_data: DataType<RichTextDefinition<string>>): string {
    return ''
  }

  get pluginControls(): RichTextPluginControl[] {
    return []
  }

  pluginControlAt(_index: number) {
    return undefined
  }
}

export { Definition as RichTextDefinition }

export const RichText = (config?: UserConfig) => new Definition(config ?? {})

RichText.Mode = Definition.Mode
