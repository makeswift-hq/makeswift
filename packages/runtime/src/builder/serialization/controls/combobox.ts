import { ComboboxControlDefinition } from '../../../controls'
import { deserializeFunction, serializeFunction } from '../function-serialization'
import { Deserialize, Serialize } from './types'

export function serializeComboboxControlDefinition(
  definition: ComboboxControlDefinition,
): [Serialize<ComboboxControlDefinition>, Transferable[]] {
  const transferables = []
  const getOptions = definition.config.getOptions && serializeFunction(definition.config.getOptions)

  if (getOptions) transferables.push(getOptions)

  return [{ ...definition, config: { ...definition.config, getOptions } }, transferables] as [
    Serialize<ComboboxControlDefinition>,
    Transferable[],
  ]
}

export function deserializeComboboxControlDefinition(
  definition: Serialize<ComboboxControlDefinition>,
): Deserialize<Serialize<ComboboxControlDefinition>> {
  const getOptions =
    definition.config.getOptions && deserializeFunction(definition.config.getOptions)

  return {
    ...definition,
    config: { ...definition.config, getOptions },
  } as Deserialize<Serialize<ComboboxControlDefinition>>
}
