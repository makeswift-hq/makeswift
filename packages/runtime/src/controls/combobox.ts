import { Data } from '@makeswift/controls'

export const ComboboxControlType = 'makeswift::controls::combobox'

type ComboboxControlOption<T extends Data> = { id: string; label: string; value: T }

type ComboboxControlConfig<T extends Data = Data> = {
  label?: string
  getOptions(query: string): ComboboxControlOption<T>[] | Promise<ComboboxControlOption<T>[]>
}

export type ComboboxControlDefinition<C extends ComboboxControlConfig = ComboboxControlConfig> = {
  type: typeof ComboboxControlType
  config: C
}

export type ComboboxControlDefinitionOption<T extends ComboboxControlDefinition> =
  T['config'] extends ComboboxControlConfig<infer U> ? U : never

export type ComboboxControlData<T extends ComboboxControlDefinition = ComboboxControlDefinition> =
  ComboboxControlOption<ComboboxControlDefinitionOption<T>>

export function Combobox<T extends Data, C extends ComboboxControlConfig<T>>(
  config: C & {
    getOptions(query: string): ComboboxControlOption<T>[] | Promise<ComboboxControlOption<T>[]>
  },
): ComboboxControlDefinition<C> {
  return { type: ComboboxControlType, config }
}
