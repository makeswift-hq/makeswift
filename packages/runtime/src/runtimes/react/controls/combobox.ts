import { type DataType, type ResolvedValueType } from '@makeswift/controls'

import { ComboboxDefinition } from '../../../controls'

export function useComboboxControlValue(
  data: DataType<ComboboxDefinition> | undefined,
): ResolvedValueType<ComboboxDefinition> {
  return data?.value
}
