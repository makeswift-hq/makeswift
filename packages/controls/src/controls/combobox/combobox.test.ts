import { testDefinition, testResolveValue } from '../../testing/test-definition'

import { type ResolvedValueType, type ValueType } from '../associated-types'

import { Combobox, ComboboxDefinition } from './combobox'

describe('Combobox', () => {
  describe('constructor', () => {
    test("definition's config type is derived from constructor's arguments", () => {
      // Assert
      Combobox({
        label: 'Nato',
        getOptions: () => [
          { id: 'a', value: 'a', label: 'alpha' },
          { id: 'b', value: 'b', label: 'beta' },
        ],
      }).config satisfies {
        label?: string
        getOptions: () => { id: string; value: 'a' | 'b'; label: string }[]
      }

      Combobox({
        label: 'Nato',
        getOptions: async () => [
          { id: 'a', value: 'a', label: 'alpha' },
          { id: 'b', value: 'b', label: 'beta' },
        ],
      }).config satisfies {
        label?: string
        getOptions: () => Promise<
          { id: string; value: 'a' | 'b'; label: string }[]
        >
      }
    })

    test('refines value type and resolved value based on options', () => {
      const comboDef = Combobox({
        label: 'Block type',
        getOptions: () => [
          { id: 'p', value: 'p', label: 'Paragraph' },
          { id: 'h1', value: 'h1', label: 'Heading 1' },
        ],
      })

      const value:
        | { id: string; value: 'p'; label: string }
        | { id: string; value: 'h1'; label: string } = {
        id: 'p',
        value: 'p',
        label: 'Paragraph',
      } as ValueType<typeof comboDef>

      const resolvedValue: 'p' | 'h1' | undefined = comboDef
        .resolveValue({ id: 'foo', value: 'p', label: 'label' })
        .readStableValue() as ResolvedValueType<typeof comboDef>

      expect(comboDef).toMatchSnapshot()
      expect(value).toMatchSnapshot()
      expect(resolvedValue).toMatchSnapshot()
    })

    test('disallows extraneous properties', () => {
      Combobox({
        label: undefined,
        getOptions: () => [],
        // @ts-expect-error
        extra: 'extra',
      })
    })
  })

  describe('assignability', () => {
    function assignTest(_def: ComboboxDefinition) {}
    const getOptions = () => [
      {
        id: '1234567890',
        value: { latitude: 41.661129, longitude: -91.530167 },
        label: 'Iowa City',
      },
    ]

    assignTest(Combobox({ label: 'location', getOptions }))
    assignTest(Combobox({ label: undefined, getOptions }))
  })
})

describe.each([
  [
    Combobox({
      label: 'Block type',
      getOptions: () => [
        { id: 'p', value: 'p', label: 'Paragraph' },
        { id: 'h1', value: 'h1', label: 'Heading 1' },
      ],
    }),
    [
      { id: 'p', value: 'p', label: 'Paragraph' },
      { id: 'h1', value: 'h1', label: 'Paragraph' },
    ] as const,
  ],
])('Combobox', (def, values) => {
  const invalidValues = [null, 17, 'random', { swatchId: 42 }]
  testDefinition(def, values, invalidValues)
  testResolveValue(def, values)
})
