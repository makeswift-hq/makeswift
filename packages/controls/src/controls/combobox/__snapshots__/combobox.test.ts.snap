// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Combobox constructor refines value type and resolved value based on options 1`] = `
ComboboxDefinition {
  "config": {
    "getOptions": [Function],
    "label": "Block type",
  },
}
`;

exports[`Combobox constructor refines value type and resolved value based on options 2`] = `
{
  "id": "p",
  "label": "Paragraph",
  "value": "p",
}
`;

exports[`Combobox constructor refines value type and resolved value based on options 3`] = `"p"`;

exports[`Combobox definition w/ config {"label":"Block type"} toData returns v1 data for \`{ id: 'h1', value: 'h1', label: 'Paragraph' }\` when definition version is 1 1`] = `
{
  "id": "h1",
  "label": "Paragraph",
  "value": "h1",
}
`;

exports[`Combobox definition w/ config {"label":"Block type"} toData returns v1 data for \`{ id: 'p', value: 'p', label: 'Paragraph' }\` when definition version is 1 1`] = `
{
  "id": "p",
  "label": "Paragraph",
  "value": "p",
}
`;

exports[`Combobox resolveValue w/ config {"label":"Block type"} resolves data ({ id: 'h1', value: 'h1', label: 'Paragraph' }) 1`] = `"h1"`;

exports[`Combobox resolveValue w/ config {"label":"Block type"} resolves data ({ id: 'p', value: 'p', label: 'Paragraph' }) 1`] = `"p"`;
